File: brain/src/service/domain_classifier.py
"""4-layer domain sensitivity classifier.

Determines the sensitivity level of user text to control PII scrub
intensity.  Higher sensitivity = more aggressive scrubbing.

Pipeline (short-circuits on SENSITIVE/LOCAL_ONLY):

    Layer 1: Persona override  — ``/health`` persona → SENSITIVE.
    Layer 2: Keyword signals   — health/finance/legal keywords → score.
    Layer 3: Vault context     — sensitive source data → SENSITIVE.
    Layer 4: LLM fallback      — only if confidence < 0.5 and LLM available.

Selection: highest-confidence layer wins.

No imports from adapter/ — only domain types.
"""

from __future__ import annotations

import re
from typing import Any

import structlog

from ..domain.enums import Sensitivity
from ..domain.types import Classification

logger = structlog.get_logger(__name__)


# ---------------------------------------------------------------------------
# Layer 1: Persona → Sensitivity mapping
# ---------------------------------------------------------------------------

_PERSONA_MAP: dict[str, tuple[Sensitivity, str]] = {
    "health": (Sensitivity.SENSITIVE, "health persona override"),
    "medical": (Sensitivity.SENSITIVE, "medical persona override"),
    "financial": (Sensitivity.ELEVATED, "financial persona override"),
    "finance": (Sensitivity.ELEVATED, "financial persona override"),
    "legal": (Sensitivity.ELEVATED, "legal persona override"),
    "work": (Sensitivity.ELEVATED, "work persona override"),
    "general": (Sensitivity.GENERAL, "general persona"),
    "personal": (Sensitivity.GENERAL, "personal → general"),
    "social": (Sensitivity.GENERAL, "social → general"),
}

# ---------------------------------------------------------------------------
# Layer 2: Keyword signals → domain + sensitivity
# ---------------------------------------------------------------------------

_HEALTH_STRONG = re.compile(
    r"\b(?:diagnosis|prescription|symptom|blood\s*(?:sugar|pressure|test)|"
    r"cholesterol|A1C|biopsy|MRI|CT\s*scan|radiology|oncology|pathology|"
    r"medication|dosage|insulin|chemotherapy|surgery|hospital|clinic|"
    r"patient|medical\s*record|lab\s*result|hemoglobin|platelet|"
    r"x-ray|ultrasound|ecg|ekg)\b",
    re.IGNORECASE,
)

_HEALTH_WEAK = re.compile(
    r"\b(?:doctor|health|wellness|diet|exercise|weight|sleep|"
    r"headache|fever|cold|flu|allergy|vitamin)\b",
    re.IGNORECASE,
)

_FINANCE_STRONG = re.compile(
    r"\b(?:bank\s*account|credit\s*card|debit\s*card|loan|mortgage|"
    r"tax\s*return|salary|income|investment|portfolio|"
    r"aadhaar|pan\s*card|pan\s*number|ifsc|neft|rtgs|upi|"
    r"account\s*number|routing\s*number|swift|iban)\b",
    re.IGNORECASE,
)

_FINANCE_WEAK = re.compile(
    r"\b(?:money|payment|price|cost|budget|expense|savings|"
    r"insurance|premium|interest\s*rate)\b",
    re.IGNORECASE,
)

_LEGAL_STRONG = re.compile(
    r"\b(?:lawsuit|subpoena|deposition|court\s*order|litigation|"
    r"attorney|lawyer|legal\s*counsel|affidavit|indictment|"
    r"bail|probation|verdict|plea|custody|restraining\s*order)\b",
    re.IGNORECASE,
)


def _keyword_classify(text: str) -> Classification | None:
    """Layer 2: keyword-based classification."""
    health_strong = len(_HEALTH_STRONG.findall(text))
    health_weak = len(_HEALTH_WEAK.findall(text))
    finance_strong = len(_FINANCE_STRONG.findall(text))
    finance_weak = len(_FINANCE_WEAK.findall(text))
    legal_strong = len(_LEGAL_STRONG.findall(text))

    # Score each domain.
    health_score = health_strong * 0.3 + health_weak * 0.1
    finance_score = finance_strong * 0.3 + finance_weak * 0.1
    legal_score = legal_strong * 0.3

    best_domain = "general"
    best_score = 0.0
    best_sensitivity = Sensitivity.GENERAL

    if health_score > best_score:
        best_score = health_score
        best_domain = "health"
        best_sensitivity = Sensitivity.SENSITIVE if health_strong else Sensitivity.ELEVATED

    if finance_score > best_score:
        best_score = finance_score
        best_domain = "financial"
        best_sensitivity = Sensitivity.SENSITIVE if finance_strong else Sensitivity.ELEVATED

    if legal_score > best_score:
        best_score = legal_score
        best_domain = "legal"
        best_sensitivity = Sensitivity.SENSITIVE

    if best_score < 0.1:
        return None

    confidence = min(best_score, 1.0)
    return Classification(
        sensitivity=best_sensitivity,
        domain=best_domain,
        reason=f"keyword signals: {best_domain} (score={best_score:.2f})",
        confidence=confidence,
    )


class DomainClassifier:
    """4-layer sensitivity classifier.

    Parameters
    ----------
    llm:
        Optional LLM client for Layer 4 fallback.  If None, Layer 4
        is skipped and low-confidence results default to GENERAL.
    """

    # Tier → Sensitivity mapping (Brain-side logic).
    _TIER_SENSITIVITY = {
        "sensitive": Sensitivity.SENSITIVE,
        "locked": Sensitivity.SENSITIVE,
        "standard": Sensitivity.ELEVATED,
        "default": Sensitivity.GENERAL,
    }

    def __init__(self, llm: Any = None, registry: Any = None) -> None:
        self._llm = llm
        self._registry = registry

    def _resolve_persona_sensitivity(self, key: str) -> tuple[Sensitivity | None, str]:
        """Resolve persona name to sensitivity level.

        Uses registry (dynamic) first, falls back to static _PERSONA_MAP.
        Returns (sensitivity, reason) or (None, "") if unknown.
        """
        if self._registry:
            tier = self._registry.tier(key)
            if tier:
                sensitivity = self._TIER_SENSITIVITY.get(tier, Sensitivity.GENERAL)
                return sensitivity, f"{key} persona (tier={tier})"
        # Fallback to static map
        if key in _PERSONA_MAP:
            return _PERSONA_MAP[key]
        return None, ""

    def classify(
        self,
        text: str,
        persona: str | None = None,
        vault_context: dict[str, Any] | None = None,
    ) -> Classification:
        """Classify text sensitivity through the 4-layer pipeline.

        Parameters
        ----------
        text:
            The user's query or message text.
        persona:
            Active persona name (e.g. "health", "financial").
        vault_context:
            Optional metadata about the source data.

        Returns
        -------
        Classification
            Sensitivity level, domain, reason, and confidence.
        """
        # Layer 1: Persona override.
        if persona:
            key = persona.strip("/").lower()
            sensitivity, reason = self._resolve_persona_sensitivity(key)
            if sensitivity is not None:
                result = Classification(
                    sensitivity=sensitivity,
                    domain=key if key in ("health", "medical", "financial", "finance", "legal") else "general",
                    reason=reason,
                    confidence=0.95,
                )
                logger.debug("domain_classifier.persona_override", result=result)
                # Short-circuit on SENSITIVE/LOCAL_ONLY.
                if result.sensitivity in (Sensitivity.SENSITIVE, Sensitivity.LOCAL_ONLY):
                    return result
                # Otherwise, continue to see if keywords push higher.
                persona_result = result
            else:
                persona_result = None
        else:
            persona_result = None

        # Layer 2: Keyword signals.
        keyword_result = _keyword_classify(text)

        # Layer 3: Vault context.
        vault_result = None
        if vault_context:
            source = vault_context.get("source", "").lower()
            item_type = vault_context.get("type", "").lower()
            if source in ("health_system", "medical", "hospital", "clinic"):
                vault_result = Classification(
                    sensitivity=Sensitivity.SENSITIVE,
                    domain="health",
                    reason=f"vault source is '{source}'",
                    confidence=0.9,
                )
            elif source in ("bank", "financial", "tax"):
                vault_result = Classification(
                    sensitivity=Sensitivity.SENSITIVE,
                    domain="financial",
                    reason=f"vault source is '{source}'",
                    confidence=0.9,
                )
            elif item_type in ("medical_record", "lab_result", "prescription"):
                vault_result = Classification(
                    sensitivity=Sensitivity.SENSITIVE,
                    domain="health",
                    reason=f"vault item type is '{item_type}'",
                    confidence=0.9,
                )

        # Select: highest confidence wins.
        candidates: list[Classification] = []
        if persona_result:
            candidates.append(persona_result)
        if keyword_result:
            candidates.append(keyword_result)
        if vault_result:
            candidates.append(vault_result)

        if candidates:
            best = max(candidates, key=lambda c: c.confidence)
            # If there's a tie in confidence, prefer higher sensitivity.
            sensitivity_rank = {
                Sensitivity.LOCAL_ONLY: 4,
                Sensitivity.SENSITIVE: 3,
                Sensitivity.ELEVATED: 2,
                Sensitivity.GENERAL: 1,
            }
            ties = [c for c in candidates if c.confidence == best.confidence]
            if len(ties) > 1:
                best = max(ties, key=lambda c: sensitivity_rank.get(c.sensitivity, 0))
            return best

        # Layer 4: LLM fallback — skipped for now (no LLM in classifier).
        # Default to GENERAL.
        return Classification(
            sensitivity=Sensitivity.GENERAL,
            domain="general",
            reason="no signals detected — default general",
            confidence=0.3,
        )

File: apps/home-node-lite/core-server/src/brain/domain_classifier.ts
/**
 * Task 5.32 — domain sensitivity classifier.
 *
 * Controls how aggressively the PII scrubber (Phase 4j) runs by
 * labelling each piece of user text with a `Sensitivity` level:
 *
 *   GENERAL      — low-risk. Default scrub intensity.
 *   ELEVATED     — meaningful exposure if leaked (work docs, legal
 *                  notes). Redact PII + keep generic terms.
 *   SENSITIVE    — health / financial records. Aggressive redaction;
 *                  cloud LLMs blocked for some personas.
 *   LOCAL_ONLY   — never leaves the device (e.g. mental-health notes).
 *
 * **Four-layer pipeline** (highest confidence wins; ties broken by
 * higher sensitivity):
 *
 *   Layer 1: **Persona override** — `/health` → SENSITIVE,
 *            `/financial` → ELEVATED. Short-circuits when
 *            confidence=0.95 AND result is SENSITIVE/LOCAL_ONLY
 *            (nothing a later layer could add).
 *   Layer 2: **Keyword signals** — weighted domain keywords. Strong
 *            health/finance terms (ssn, diagnosis) compound toward
 *            SENSITIVE; weaker terms (doctor, payment) toward
 *            ELEVATED.
 *   Layer 3: **Vault context** — optional metadata about the source
 *            of the text. A vault item tagged `source: "hospital"`
 *            forces SENSITIVE regardless of keywords.
 *   Layer 4: **LLM fallback** — optional hook called only when the
 *            deterministic layers find nothing (confidence < 0.5).
 *            Injectable so tests can cover the layer without an
 *            LLM dependency.
 *
 * **Pure + deterministic** when no LLM is wired. Injectable clock
 * is unnecessary because no time-based decisions live here.
 *
 * Source: docs/HOME_NODE_LITE_TASKS.md Phase 5e task 5.32.
 */

export type Sensitivity = 'general' | 'elevated' | 'sensitive' | 'local_only';

export type Domain = 'general' | 'health' | 'financial' | 'legal' | 'work';

export interface Classification {
  sensitivity: Sensitivity;
  domain: Domain;
  /** Short human-readable rationale, safe for logs. */
  reason: string;
  /** 0..1; higher = more certain. */
  confidence: number;
  /** Which layer produced this classification (for telemetry + debug). */
  layer: 'persona' | 'keyword' | 'vault' | 'llm' | 'default';
}

export interface VaultContext {
  /** Origin of the vault item (e.g. `"hospital"`, `"bank"`). */
  source?: string;
  /** Type of vault item (e.g. `"medical_record"`, `"lab_result"`). */
  type?: string;
}

/** Async LLM call. Resolves with a domain hint string. */
export type DomainLlmCallFn = (text: string) => Promise<{
  domain: Domain;
  sensitivity: Sensitivity;
  reason?: string;
}>;

export interface DomainClassifierOptions {
  /** Optional LLM fallback — only called when layers 1–3 are inconclusive. */
  llmCallFn?: DomainLlmCallFn;
  /**
   * Minimum confidence that shortcuts the LLM layer. Defaults to 0.5.
   * If deterministic layers produce ≥ threshold, LLM is skipped.
   */
  llmConfidenceThreshold?: number;
  /**
   * Persona registry adapter — `tier(name)` returns the current tier
   * for a persona, or `null` if unknown. When present, tier→sensitivity
   * takes precedence over the static persona map (dynamic per-user
   * configuration beats hardcoded defaults).
   */
  personaRegistry?: { tier(persona: string): string | null };
}

export const DEFAULT_LLM_CONFIDENCE_THRESHOLD = 0.5;

// ── Static persona → (sensitivity, domain) map ─────────────────────────

const PERSONA_MAP: ReadonlyMap<
  string,
  { sensitivity: Sensitivity; domain: Domain; reason: string }
> = new Map([
  ['health', { sensitivity: 'sensitive', domain: 'health', reason: 'health persona override' }],
  ['medical', { sensitivity: 'sensitive', domain: 'health', reason: 'medical persona override' }],
  ['financial', { sensitivity: 'elevated', domain: 'financial', reason: 'financial persona override' }],
  ['finance', { sensitivity: 'elevated', domain: 'financial', reason: 'financial persona override' }],
  ['legal', { sensitivity: 'elevated', domain: 'legal', reason: 'legal persona override' }],
  ['work', { sensitivity: 'elevated', domain: 'work', reason: 'work persona override' }],
  ['general', { sensitivity: 'general', domain: 'general', reason: 'general persona' }],
  ['personal', { sensitivity: 'general', domain: 'general', reason: 'personal → general' }],
  ['social', { sensitivity: 'general', domain: 'general', reason: 'social → general' }],
]);

const TIER_SENSITIVITY: ReadonlyMap<string, Sensitivity> = new Map([
  ['sensitive', 'sensitive'],
  ['locked', 'sensitive'],
  ['standard', 'elevated'],
  ['default', 'general'],
]);

// ── Keyword signals ────────────────────────────────────────────────────

/** Health terms that *by themselves* warrant SENSITIVE (e.g. diagnoses). */
const HEALTH_STRONG = [
  'diagnosis', 'diagnosed', 'prescription', 'prescribed', 'hiv',
  'cancer', 'tumor', 'tumour', 'chemotherapy', 'surgery', 'lab result',
  'blood test', 'mri', 'x-ray', 'biopsy', 'depression', 'anxiety disorder',
  'medication', 'medical record', 'health insurance',
];
/** Weaker health context — contributes to ELEVATED but not SENSITIVE. */
const HEALTH_WEAK = [
  'doctor', 'hospital', 'clinic', 'nurse', 'appointment', 'checkup',
  'pharmacy', 'therapist', 'dentist', 'vaccine', 'shot', 'pain',
];
/** Financial strong terms — bank accounts, SSN, etc. */
const FINANCE_STRONG = [
  'ssn', 'social security', 'account number', 'routing number',
  'credit card', 'debit card', 'cvv', 'pin', 'net worth',
  'bank balance', 'tax return', 'w-2', '1099',
];
/** Weaker financial context. */
const FINANCE_WEAK = [
  'bank', 'payment', 'invoice', 'mortgage', 'loan', 'investment',
  'portfolio', 'broker', 'dividend', 'salary', 'paycheck',
];
/** Legal-domain terms — contracts, litigation. */
const LEGAL_STRONG = [
  'lawsuit', 'litigation', 'plaintiff', 'defendant', 'subpoena',
  'deposition', 'settlement', 'contract', 'nda', 'indictment',
  'custody', 'divorce decree', 'estate planning',
];

/**
 * Escape a string so it can be embedded in a regex literal. The
 * keyword lists contain dots, dashes, and spaces — `w-2`, `social
 * security`, `x-ray` — so we can't just concatenate into a regex.
 */
function escapeRegex(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

/**
 * Count word-boundary matches of each keyword in `text`. Using word
 * boundaries matters: "pin" is in FINANCE_STRONG and naive substring
 * matching would flag "opinion" + "pinpoint" + "typing" as financial
 * keywords. Whitespace, punctuation, and start/end-of-text count as
 * boundaries. Case-insensitive.
 */
function countMatches(text: string, keywords: readonly string[]): number {
  if (text.length === 0) return 0;
  let n = 0;
  for (const kw of keywords) {
    if (kw.length === 0) continue;
    // Build a case-insensitive regex with (?:^|\W) and (?:\W|$) as
    // boundaries. We can't use \b because some keywords contain
    // dashes (`x-ray`, `w-2`) — \b after a dash wouldn't match.
    const re = new RegExp(`(?:^|\\W)${escapeRegex(kw)}(?:\\W|$)`, 'gi');
    const matches = text.match(re);
    if (matches) n += matches.length;
  }
  return n;
}

function keywordClassify(text: string): Classification | null {
  const healthStrong = countMatches(text, HEALTH_STRONG);
  const healthWeak = countMatches(text, HEALTH_WEAK);
  const financeStrong = countMatches(text, FINANCE_STRONG);
  const financeWeak = countMatches(text, FINANCE_WEAK);
  const legalStrong = countMatches(text, LEGAL_STRONG);

  const healthScore = healthStrong * 0.3 + healthWeak * 0.1;
  const financeScore = financeStrong * 0.3 + financeWeak * 0.1;
  const legalScore = legalStrong * 0.3;

  let best: { score: number; domain: Domain; sensitivity: Sensitivity } = {
    score: 0,
    domain: 'general',
    sensitivity: 'general',
  };
  if (healthScore > best.score) {
    best = {
      score: healthScore,
      domain: 'health',
      sensitivity: healthStrong > 0 ? 'sensitive' : 'elevated',
    };
  }
  if (financeScore > best.score) {
    best = {
      score: financeScore,
      domain: 'financial',
      sensitivity: financeStrong > 0 ? 'sensitive' : 'elevated',
    };
  }
  if (legalScore > best.score) {
    best = { score: legalScore, domain: 'legal', sensitivity: 'sensitive' };
  }

  if (best.score < 0.1) return null;

  const confidence = Math.min(best.score, 1.0);
  return {
    sensitivity: best.sensitivity,
    domain: best.domain,
    reason: `keyword signals: ${best.domain} (score=${best.score.toFixed(2)})`,
    confidence,
    layer: 'keyword',
  };
}

function vaultContextClassify(ctx: VaultContext): Classification | null {
  const source = (ctx.source ?? '').toLowerCase();
  const type = (ctx.type ?? '').toLowerCase();
  const HEALTH_SOURCES = new Set(['health_system', 'medical', 'hospital', 'clinic']);
  const FINANCE_SOURCES = new Set(['bank', 'financial', 'tax']);
  const HEALTH_TYPES = new Set(['medical_record', 'lab_result', 'prescription']);

  if (HEALTH_SOURCES.has(source)) {
    return {
      sensitivity: 'sensitive',
      domain: 'health',
      reason: `vault source is "${source}"`,
      confidence: 0.9,
      layer: 'vault',
    };
  }
  if (FINANCE_SOURCES.has(source)) {
    return {
      sensitivity: 'sensitive',
      domain: 'financial',
      reason: `vault source is "${source}"`,
      confidence: 0.9,
      layer: 'vault',
    };
  }
  if (HEALTH_TYPES.has(type)) {
    return {
      sensitivity: 'sensitive',
      domain: 'health',
      reason: `vault item type is "${type}"`,
      confidence: 0.9,
      layer: 'vault',
    };
  }
  return null;
}

const SENSITIVITY_RANK: Readonly<Record<Sensitivity, number>> = {
  local_only: 4,
  sensitive: 3,
  elevated: 2,
  general: 1,
};

export class DomainClassifier {
  private readonly llmCallFn?: DomainLlmCallFn;
  private readonly llmConfidenceThreshold: number;
  private readonly personaRegistry?: { tier(persona: string): string | null };

  constructor(opts: DomainClassifierOptions = {}) {
    if (opts.llmCallFn !== undefined) this.llmCallFn = opts.llmCallFn;
    this.llmConfidenceThreshold =
      opts.llmConfidenceThreshold ?? DEFAULT_LLM_CONFIDENCE_THRESHOLD;
    if (opts.personaRegistry !== undefined) this.personaRegistry = opts.personaRegistry;
  }

  /**
   * Classify `text` with optional `persona` + `vaultContext`. Never
   * throws. When the LLM fallback is wired, it's only called if the
   * deterministic layers produce a result with confidence below
   * `llmConfidenceThreshold`.
   */
  async classify(input: {
    text: string;
    persona?: string;
    vaultContext?: VaultContext;
  }): Promise<Classification> {
    const text = input.text ?? '';
    const candidates: Classification[] = [];

    // Layer 1: persona override.
    const personaResult = this.classifyByPersona(input.persona);
    if (personaResult) {
      candidates.push(personaResult);
      // Short-circuit on top-tier sensitivity — nothing else could raise it.
      if (
        personaResult.sensitivity === 'sensitive' ||
        personaResult.sensitivity === 'local_only'
      ) {
        return personaResult;
      }
    }

    // Layer 2: keywords.
    const keywordResult = keywordClassify(text);
    if (keywordResult) candidates.push(keywordResult);

    // Layer 3: vault context.
    if (input.vaultContext) {
      const vaultResult = vaultContextClassify(input.vaultContext);
      if (vaultResult) candidates.push(vaultResult);
    }

    const best = pickBest(candidates);
    if (best && best.confidence >= this.llmConfidenceThreshold) {
      return best;
    }

    // Layer 4: LLM fallback (optional).
    if (this.llmCallFn) {
      try {
        const raw = await this.llmCallFn(text);
        const coerced = coerceLlmResult(raw);
        if (coerced) {
          return coerced;
        }
      } catch {
        // LLM failure is non-fatal — fall through to best-or-default.
      }
    }

    return (
      best ?? {
        sensitivity: 'general',
        domain: 'general',
        reason: 'no signals detected — default general',
        confidence: 0.3,
        layer: 'default',
      }
    );
  }

  private classifyByPersona(persona: string | undefined): Classification | null {
    if (typeof persona !== 'string' || persona.length === 0) return null;
    const key = persona.trim().replace(/^\/+/, '').toLowerCase();
    if (key === '') return null;

    // Registry wins when present (dynamic per-user config).
    if (this.personaRegistry) {
      const tier = this.personaRegistry.tier(key);
      if (tier !== null) {
        const sensitivity = TIER_SENSITIVITY.get(tier) ?? 'general';
        return {
          sensitivity,
          domain: PERSONA_MAP.get(key)?.domain ?? 'general',
          reason: `${key} persona (tier=${tier})`,
          confidence: 0.95,
          layer: 'persona',
        };
      }
    }
    // Fallback to static map.
    const mapped = PERSONA_MAP.get(key);
    if (mapped) {
      return {
        sensitivity: mapped.sensitivity,
        domain: mapped.domain,
        reason: mapped.reason,
        confidence: 0.95,
        layer: 'persona',
      };
    }
    return null;
  }
}

// ── Internals ──────────────────────────────────────────────────────────

function pickBest(candidates: Classification[]): Classification | null {
  if (candidates.length === 0) return null;
  let best = candidates[0]!;
  for (const c of candidates.slice(1)) {
    if (c.confidence > best.confidence) best = c;
    else if (
      c.confidence === best.confidence &&
      SENSITIVITY_RANK[c.sensitivity] > SENSITIVITY_RANK[best.sensitivity]
    ) {
      best = c;
    }
  }
  return best;
}

const ALLOWED_SENSITIVITY: ReadonlySet<Sensitivity> = new Set([
  'general', 'elevated', 'sensitive', 'local_only',
]);
const ALLOWED_DOMAIN: ReadonlySet<Domain> = new Set([
  'general', 'health', 'financial', 'legal', 'work',
]);

function coerceLlmResult(raw: unknown): Classification | null {
  if (raw === null || typeof raw !== 'object') return null;
  const r = raw as Partial<{ domain: Domain; sensitivity: Sensitivity; reason: string }>;
  if (!r.sensitivity || !ALLOWED_SENSITIVITY.has(r.sensitivity)) return null;
  if (!r.domain || !ALLOWED_DOMAIN.has(r.domain)) return null;
  return {
    sensitivity: r.sensitivity,
    domain: r.domain,
    reason: typeof r.reason === 'string' && r.reason.length > 0 ? r.reason : 'LLM classification',
    confidence: 0.6,
    layer: 'llm',
  };
}
